### ISSUE TITLE
Real-World  QML Algorithms

### ISSUE THREAD

**Issue by ElePT** [qiskit-community/mentorship-program-fall-21#43]
# Description
The goal of this project is to expand our current understanding of the different **practical aspects** of recently proposed QML algorithms (i.e [[1]](https://arxiv.org/abs/1907.02085), [[2]](https://arxiv.org/abs/2103.14653), [[3]](https://arxiv.org/abs/1912.08278)). These could include factors such as computational cost, execution time, level of applicability to real-world scenarios and other challenges related to software implementation or execution in current quantum devices. 

The main idea is to  implement one or more recent QML algorithm proposals in Qiskit, and benchmark them against the already existing qiskit-machine-learning algorithms (such as VQC or QKE). Furthermore, these algorithms could be tested in one or more industrial applications to  help showcase the potential contributions of QML to different fields. 

The scope of the project can be adjusted to fit within the expected  3-month timeline.

# Mentor/s
Matched by prior communication with Alex Pozas-Kerstjens @apozas (should he accept the challenge)

# Type of participant
I accept this challenge :)

# Number of participants
1

# Deliverable
- (1+) PRs on the qiskit-machine-learning repo for the newly implemented QML algorithms
- Report/ poster/ potential publication summarizing our findings

**Comment by apozas** [comment]
I accept the challenge

**Comment by ElePT** [comment]
Here are the slides for the intermediate presentation:
[#43 Real-World QML Algorithms.pdf](https://github.com/qiskit-advocate/qamp-fall-21/files/7293095/43.Real-World.QML.Algorithms.pdf)

**Comment by ElePT** [comment]
And here is the update for **checkpoint 2**:

Because the original project definition was quite open, for checkpoint 2 we have focused on developing a more concrete approach to benchmarking QML algorithms, starting with Quantum Neural Networks. 

On one hand, we have selected **two datasets** to use throughout the project:
1.	The first dataset is a simple 2D classification dataset, where points are either in or out of a circle. This dataset is very simple and can be used to verify that the implementations work. 
2.	The second dataset is from Kaggle and defines the actual real-world task: credit card fraud classification. Originally, this is a binary classification dataset with 30 features, but the number of features can be adjusted to the current limitations in number of qubits in a pre-processing step (as it is commonly done in data science problems).

On the other hand, we have established a **criterion for classical baselines**: because it is not “fair” to compare a deep neural network to a basic quantum machine learning algorithm, we have determined that a “comparable” classical neural network could be one with a similar number of trainable parameters. 
 
In terms of algorithm implementation, we have decided to follow a **bottom-up strategy**, starting from the most fundamental algorithm, and building up in number of qubits and gates. According to this logic, the first algorithm to implement was the single-qubit universal quantum classifier[ [1]](https://arxiv.org/abs/1907.02085) [[2]](https://arxiv.org/pdf/2102.04032.pdf), a variational quantum circuit with 3 alternative theoretical formulations and trained through a fidelity loss function. To implement this fidelity loss effectively, we had to adapt the `BinaryObjectiveFunction `class from `qiskit-machine-learning`. 

After some preliminary tests using the circle dataset, we selected the best formulations, trained the networks using the credit card fraud detection data, and compared the results to the previously established classical baselines: 
 
![image](https://user-images.githubusercontent.com/57907331/141319443-6eff9bf3-8815-4904-b5bb-7187d92caf22.png)

We have seen that the best single-qubit approach can achieve an **accuracy comparable to the classical baseline** in this real-world data, and this is really promising. In the following steps, we will explore how can entanglement and more complex variational architectures help improve these results. If necessary, we can make the task more challenging by modifying the pre-processing step.

**Comment by ElePT** [comment]
As for the image, here is the visual comparison of the quantum and classical networks on the credit card dataset:

![image](https://user-images.githubusercontent.com/57907331/141319668-7b0ca6b4-ecba-4216-97b5-a474a69f7cf7.png)

**Comment by ElePT** [comment]
The final slides are the following:
[Final_ppt_issue_43.pdf](https://github.com/qiskit-advocate/qamp-fall-21/files/7680952/Final_ppt_issue_43.pdf)
